### DISCUSSION TITLE
Asking help for getting the inserted stroke (through plugin)  selected and allowUndoRedoActon working properly

### DISCUSSION THREAD

**Discussion by MiltonBalaOfficial** [xournalpp/xournalpp#5999] [Q&A]
I have made a plugin through which I can able to insert different kinds of geometrical shapes. But I am unable to make the inserted strokes selected after it is added (Actually after copy/paste the stroke remains selected, I want like this). I also cannot make the allowUndoRedoAction working! Somebody please guide me.

My codes are like that: 

`-- Register Toolbar actions and initialize UI stuff
      function initUi()
        app.registerUi({["menu"] = "Insert Shapes", ["callback"] = "geometricTwoDimensional", ["accelerator"] = "<Control><Alt>3", ["toolbarId"] = "shapedialog", ["iconName"] = "shapes_symbolic"})   
      end

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-- Calculation for finding centre of page if needed ahead
    local pageWidth = 907.08661 -- app.getDocument doesnt work globally, so assume a 16 cm * 32 cm page
    local pageHeight = 510.23622
    local centerX = pageWidth / 2
    local centerY = pageHeight / 2

-- Callback for the Shape Dialog (Geometric Two Dimensional Shapes)

        function geometricTwoDimensional()
          app.openDialog("Geometric Two Dimensional Shapes", {"Arrowhead", "Triangle", "Equilateral Triangle", "Right Angle Triangle", "Cancel" }, "geometricTwoDimensionalCallback")
        end

            
               function geometricTwoDimensionalCallback(result)
                  if result == 1 then
                       qshape_Arrowhead()
                     elseif result == 2 then
                       qshape_Triangle()
                      elseif result == 3 then
                        qshape_Equi_Triangle()
                      elseif result == 4 then
                        qshape_RightAngleTriangle()
                  end
               end

  
    -- All Strokes for Geometric-2D Shapes

        -- Callback for inserting arrowhead
            function qshape_Arrowhead()
              local arrowhead = {
                x = { centerX - 5, centerX + 5, centerX - 5 },
                y = { centerY - 10, centerY - 5, centerY },
                width = 2.26, 
                tool = "pen", 
                color = 0xff0000,
                capStyle = "round",
                lineStyle = "solid",
              }
            
              app.addStrokes { strokes = { arrowhead } }
              app.refreshPage()
            end
    
        -- Callback for inserting triangle
            function qshape_Triangle()
              local height = math.sqrt(3) * 50  -- Height calculation for an equilateral triangle with a base of 100
            
              local equilateralTriangle = { 
                x = { centerX - 50, centerX + 50, centerX + 100, centerX - 50 },  -- Repeating the first point to close the triangle
                y = { centerY + (height / 2), centerY + (height / 2), centerY - (height / 2), centerY + (height / 2) },  -- Adjusted y-coordinates for height
                width = 2.26, 
                tool = "pen", 
                color = 0xff0000,
                lineStyle = "solid",
              }
              app.addStrokes { strokes = { equilateralTriangle } }
              app.refreshPage()
            end
            
        -- Callback for inserting triangle
            function qshape_Equi_Triangle()
              local height = math.sqrt(3) * 50  -- Height calculation for an equilateral triangle with a base of 100
            
              local equilateralTriangle = { 
                x = { centerX - 50, centerX + 50, centerX, centerX - 50 },  -- Repeating the first point to close the triangle
                y = { centerY + (height / 2), centerY + (height / 2), centerY - (height / 2), centerY + (height / 2) },  -- Adjusted y-coordinates for height
                width = 2.26, 
                tool = "pen", 
                color = 0xff0000,
                capStyle = "round",
                lineStyle = "solid",
              }

                app.addStrokes { strokes = { equilateralTriangle } }
                app.refreshPage()
            end      

        -- Callback for inserting right-angle isosceles triangle
            function qshape_RightAngleTriangle()
              local rightAngleTriangle = {
                x = { centerX + 50, centerX - 50, centerX - 50, centerX + 50 },  
                y = { centerY + 50, centerY + 50, centerY - 50, centerY + 50 },  
                width = 2.26, 
                tool = "pen", 
                color = 0x0000ff,  
                capStyle = "round",
                lineStyle = "solid",
              }
            
              app.addStrokes { strokes = { rightAngleTriangle } }
              app.refreshPage()
            end`

**Comment by rolandlo** [comment]
Currently the strokes are simply inserted into the document, not into a selection and there is no way to create a selection out of it via a plugin. You can make a feature request about it on the issue tracker.

What concerns "allowUndoRedoAction": For this you can specify "grouped", "individual" or "none". The first of those is about the case, where you add multiple strokes at once, which you don't do.

**Reply by MiltonBalaOfficial** [reply]
Thanks, it will be very helpfull if you kindly provide an example (code) regarding "allowUndoRedoAction".

**Reply by rolandlo** [reply]
Here is an example:
```lua
triangle = { 
  x = { 300.0, 500.0, 400.0 }, 
  y = { 300.0, 300.0, 450.0 }, 
  width = 1.5, 
  tool = "pen", 
  color = 0xff0000,
}
square = { 
  x = { 300.0, 500.0, 500.0, 300.0 }, 
  y = { 500.0, 500.0, 700.0, 700.0 }, 
  width = 1.5, 
  tool = "pen", 
  color = 0x00ff00,
}
app.addStrokes { strokes = { triangle, square }, allowUndoRedoAction = "grouped" }
app.refreshPage()
```
Replace "grouped" by 
- "individual" if undo should first undo the square and then the triangle (not both at once)
- "none" if no undo action should be created.

**Reply by MiltonBalaOfficial** [reply]
I tried in this way before asking but not worked, thought that I must have done mistake somewhere. I have  also tried with your code, but not working. But I noticed that if I add stroke through this code and want to close the app, it asks if I want to save the file. But if I press CTRL+Z after adding the stroke (though strokes remain) and then closes the application, the app closes without asking anything! I cannot figure out what wrong I am doing. Any suggestion for me?

`      function initUi()
        app.registerUi({["menu"] = "Check Undo Redo", ["callback"] = "CheckUndoRedo"})   
      end
      
      function CheckUndoRedo()
        triangle = { 
          x = { 300.0, 500.0, 400.0 }, 
          y = { 300.0, 300.0, 450.0 }, 
          width = 1.5, 
          tool = "pen", 
          color = 0xff0000,
        }
        square = { 
          x = { 300.0, 500.0, 500.0, 300.0 }, 
          y = { 500.0, 500.0, 700.0, 700.0 }, 
          width = 1.5, 
          tool = "pen", 
          color = 0x00ff00,
        }
        app.addStrokes { strokes = { triangle, square }, allowUndoRedoAction = "grouped" }
        app.refreshPage()
      end`

**Reply by rolandlo** [reply]
I have tried now as well. This looks like a regression. I will investigate.

**Reply by rolandlo** [reply]
I have opened issue https://github.com/xournalpp/xournalpp/issues/6015 about it.

**Reply by rolandlo** [reply]
This is fixed in master now and the fix will be part of the next nightly release.
